File: 1001~1500/1220. Count Vowels Permutation/main.py
"""
Source: https://rb.gy/orxun
Date: 2023/8/21
Skill:
Ref:
Runtime: 843 ms, faster than 19.54%
Memory Usage: 184.97 MB, less than 7.33%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right
from functools import cache


class Solution:
    def countVowelPermutation(self, n: int) -> int:
        mod, relation = 10 ** 9 + 7, {(0, 1), (1, 0), (1, 2), (2, 0), (2, 1), (2, 3), (2, 4), (3, 2), (3, 4), (4, 0)}

        @cache
        def dfs(prev, length):
            if length == 0:
                return 1
            num = 0
            for i in range(5):
                if (prev, i) in relation:
                    num += dfs(i, length - 1) % mod
                    num %= mod
            return num % mod

        res = 0
        for i in range(5):
            res += dfs(i, n - 1) % mod
            res %= mod
        return res


if __name__ == "__main__":
    s = Solution()

File: 1001~1500/1220. Count Vowels Permutation/main.cpp
/**
 * Source: tinyurl.com/yusdgsyd
 * Date: 2023/10/28
 * Skill:
 * Runtime: 0 ms, faster than 100.00% of C++ online submissions
 * Memory Usage: 6.19 MB, less than 93.28% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>

using namespace std;

#define ll long long

class Solution {
public:
    int countVowelPermutation(int n) {
        ll mod = 1e9+7, res = 0;
        ll dp[5], dp2[5];
        for (int i = 0; i < 5; i++) dp[i] = 1;
        for (int i = 2; i <= n; i++) {
            for (int i = 0; i < 5; i++) dp2[i] = dp[i];
            dp[0] = (dp2[1] + dp2[2] + dp2[4]) % mod;
            dp[1] = (dp2[0] + dp2[2]) % mod;
            dp[2] = (dp2[1] + dp2[3]) % mod;
            dp[3] = dp2[2] % mod;
            dp[4] = (dp2[2] + dp2[3]) % mod;
        }
        for (int i = 0; i < 5; i++) res = (res + dp[i]) % mod;
        return res;
    }
};


int main() {
//    Solution s;
}